unstructured_ingest/v2/processes/connectors/sql/vastdb.py

```diff
@@ -0,0 +1,281 @@
+    def precheck(self) -> None:
+        try:
+            with self.get_cursor() as cursor:
+                table = self.connection_config.get_table(cursor, self.index_config.table_name)
+                table.select()
+        except Exception as e:
+            logger.error(f"failed to validate connection: {e}", exc_info=True)
+            raise DestinationConnectionError(f"failed to validate connection: {e}")
+
+
+class VastdbDownloaderConfig(SQLDownloaderConfig):
+    pass
+
+
+@dataclass
+class VastdbDownloader(SQLDownloader):
+    connection_config: VastdbConnectionConfig
+    download_config: VastdbDownloaderConfig
+    connector_type: str = CONNECTOR_TYPE
+
+    @requires_dependencies(["vastdb", "ibis", "pyarrow"], extras="vastdb")
+    def query_db(self, file_data: SqlBatchFileData) -> tuple[list[tuple], list[str]]:
+        from ibis import _  # imports the Ibis deferred expression
+
+        table_name = file_data.additional_metadata.table_name
+        id_column = file_data.additional_metadata.id_column
+        ids = tuple([item.identifier for item in file_data.batch_items])
+
+        with self.connection_config.get_cursor() as cursor:
+            table = self.connection_config.get_table(cursor, table_name)
+
+            predicate = _[id_column].isin(ids)
+
+            if self.download_config.fields:
+                # Vastdb requires the id column to be included in the fields
+                fields = self.download_config.fields + [id_column]
+                # dict.fromkeys to remove duplicates and keep order
+                reader = table.select(columns=list(dict.fromkeys(fields)), predicate=predicate)
+            else:
+                reader = table.select(predicate=predicate)
+            results = reader.read_all()
+            df = results.to_pandas()
+            return [tuple(r) for r in df.to_numpy()], results.column_names
+
+
+class VastdbUploadStagerConfig(SQLUploadStagerConfig):
+    rename_columns_map: Optional[dict] = None
+    additional_columns: Optional[list[str]] = field(default_factory=list)
+
+
+class VastdbUploadStager(SQLUploadStager):
+    upload_stager_config: VastdbUploadStagerConfig
+
+    def conform_dict(self, element_dict: dict, file_data: FileData) -> dict:
+        data = element_dict.copy()
+        metadata: dict[str, Any] = data.pop("metadata", {})
+        data_source = metadata.pop("data_source", {})
+        coordinates = metadata.pop("coordinates", {})
+
+        data.update(metadata)
+        data.update(data_source)
+        data.update(coordinates)
+
+        data["id"] = get_enhanced_element_id(element_dict=data, file_data=file_data)
+
+        # remove extraneous, not supported columns
+        # but also allow for additional columns
+        approved_columns = set(list(_COLUMNS) + self.upload_stager_config.additional_columns)
+        element = {k: v for k, v in data.items() if k in approved_columns}
+        element[RECORD_ID_LABEL] = file_data.identifier
+        return element
```

Comment:
Most of this code can be reused from the parent class implementation. Looks like this only adds the use if the additional colunmns: 
```suggestion
    def conform_dict(self, element_dict: dict, file_data: FileData) -> dict:
        conformed_dict = super().conform_dict(element_dict=element_dict, file_data=file_data)
        additional_keys = {k: v for k, v in element_dict.items() if k in self.upload_stager_config.additional_columns}
        conformed_dict.update(additional_keys)
        return conformed_dict
```

Reply:
I like it but we would still have to "flatten" element_dict again since the metadata needs to be flattened.

I will probably keep it. Unless we want to make the flattening a whole different method... in the sql file